*: Assorted fixes to get e2e-aws working again (Calico -> Flannel, etc.)

wking opened the pull request:
Patterned on the existing [`worker_ingress_kubelet_insecure_from_master`][0] from b620c16f (coreos/tectonic-installer#264).

This should address errors like:

```console
$ curl -s https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/150/pull-ci-origin-installer-e2e-aws/546/artifacts/e2e-aws/nodes/ip-10-0-52-134.ec2.internal/journal.gz | zcat >journal
$ journalread journal | grep 'current config label' | head -n3
2018-08-20T20:30:08.000827895Z I0820 20:30:08.826840       1 tnc.go:375] Node ip-10-0-134-147.ec2.internal does not have a current config label
2018-08-20T20:30:08.00082814Z  I0820 20:30:08.826860       1 tnc.go:375] Node ip-10-0-153-195.ec2.internal does not have a current config label
2018-08-20T20:30:08.000828371Z I0820 20:30:08.826866       1 tnc.go:375] Node ip-10-0-166-239.ec2.internal does not have a current config label
```

on the master node:

```console
$ journalread journal | grep -A15 'Starting Ignition' | grep -v INFO
2018-08-20T20:21:40.00097323Z  Starting Ignition (files)...
2018-08-20T20:21:40.000991225Z DEBUG    : parsed url from cmdline: ""
2018-08-20T20:21:41.000010266Z DEBUG    : parsing config: {
2018-08-20T20:21:41.0000122Z   "ignition": {
2018-08-20T20:21:41.000014165Z "config": {
2018-08-20T20:21:41.000016186Z "append": [
2018-08-20T20:21:41.000018193Z {
2018-08-20T20:21:41.000023296Z "source": "http://ci-op-imi5mbig-68485-tnc.origin-ci-int-aws.dev.rhcloud.com:80/config/master",
```

which were resulting in:

```console
$ curl -s https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/150/pull-ci-origin-installer-e2e-aws/546/build-log.txt | grep Ginkgo
   |  Ginkgo timed out waiting for all parallel nodes to report back!  |
 Ginkgo ran 1 suite in 10m6.626061944s
```

Inbound 10250 is [the kubelet API used by the control plane][1]. @smarterclayton suspects the e2e-aws tests are trying to get metrics from the kubelets, and hanging on the etcd kubelet because this rule was missing.  I'm not clear why we've only been seeing this issue for the last week though.

The third commit in this PR adds the new rule.  The previous two commits pivot from inline `ingress` and `egress` rules to stand-alone `aws_security_group_rule` resources, finishing a transition away from inline ingress/egress rules begun by coreos/tectonic-installer#264.  More details in the first two commit messages.

[0]: https://github.com/openshift/installer/blob/ffca3e23530bad0efb2ebaef8d5d9cb147b6374f/modules/aws/vpc/sg-worker.tf#L121-L129
[1]: https://kubernetes.io/docs/setup/independent/install-kubeadm/#worker-node-s

wking commented:
[The e2e-aws error][1] was:

```
Waiting for API at https://ci-op-ng25jj4j-68485-api.origin-ci-int-aws.dev.rhcloud.com:6443 to respond ...
...
Waiting for API at https://ci-op-ng25jj4j-68485-api.origin-ci-int-aws.dev.rhcloud.com:6443 to respond ...
Interrupted
```

I don't know if that's further along than the Ginkgo error or not, but I'll check the node logs later.

[1]: https://deck-ci.svc.ci.openshift.org/log?job=pull-ci-origin-installer-e2e-aws&id=553

wking commented:
> I don't know if that's further along than the Ginkgo error or not, but I'll check the node logs later.

Unfortunately, it looks like [job 553 failed to capture node logs][1], so I don't know if this has addressed the "does not have a current config label" issue or not.  On the off chance that the failed-log-capture was a flake, I'll try the test again:

/retest

[1]: https://gcsweb-ci.svc.ci.openshift.org/gcs/origin-ci-test/pr-logs/pull/openshift_installer/151/pull-ci-origin-installer-e2e-aws/553/artifacts/e2e-aws/

yifan-gu commented:
The test is stuck

eparis commented on modules/aws/vpc/sg-etcd.tf:
```diff
@@ -6,44 +6,64 @@ resource "aws_security_group" "etcd" {
-    self      = true
-  }
+  protocol  = "tcp"
+  from_port = 10250
```
Are all of the 'from' correct? don't most things trying to connect to here use a random high port?

wking replied:
> Are all of the 'from' correct? don't most things trying to connect to here use a random high port?

Yeah, that doesn't make sense to me either.  But it's what we have had in master since forever; see, for example, [here][1].  Still, I'll drop them and see if that helps.

[1]: https://github.com/openshift/installer/commit/b620c16f0a4b82ee65b2d248fd98512a9f7e430e#diff-952765b0711a2c98e18e383033131664R27

wking replied:
> > Are all of the 'from' correct? don't most things trying to connect to here use a random high port?
[
> Yeah, that doesn't make sense to me either...

Ah, [`from_port`][1] and [`to_port`][2] are for a range of ports on a single host, not the ports for both hosts involved in the connection.  So having `from_port` makes sense.

[1]: https://www.terraform.io/docs/providers/aws/r/security_group_rule.html#from_port-1
[2]: https://www.terraform.io/docs/providers/aws/r/security_group_rule.html#to_port-1

wking commented:
I got [the same error][1] [again][2] with build 555, so probably not a flake.  Still not sure what's going on there...

[1]: https://github.com/openshift/installer/pull/151#issuecomment-414524652
[2]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/151/pull-ci-origin-installer-e2e-aws/555/build-log.txt

wking commented:
I've spun off some tangential changes into #155 and #156.  I don't think they were making semantic changes, but they were at least *touching* the API ingress rule (on port 6443) and therefore might have been causing [our API timeouts][1].

[1]: https://github.com/openshift/installer/pull/151#issuecomment-414524652

wking commented:
> I've spun off some tangential changes into #155 and #156.

And we're [still][1] getting:

```
Waiting for API at https://ci-op-hpjfcvwz-68485-api.origin-ci-int-aws.dev.rhcloud.com:6443 to respond ...
...
Waiting for API at https://ci-op-hpjfcvwz-68485-api.origin-ci-int-aws.dev.rhcloud.com:6443 to respond ...
Interrupted
```

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/151/pull-ci-origin-installer-e2e-aws/557/build-log.txt

wking commented:
Looks like I may have broken the Terraform ignition setup?  [The e2e-aws test failed][1] with:

```

* module.assets_base.data.ignition_config.etcd: 3 error(s) occurred:

* module.assets_base.data.ignition_config.etcd[2]: data.ignition_config.etcd.2: unexpected EOF
* module.assets_base.data.ignition_config.etcd[0]: data.ignition_config.etcd.0: unexpected EOF
* module.assets_base.data.ignition_config.etcd[1]: data.ignition_config.etcd.1: unexpected EOF
* module.assets_base.module.bootkube.data.ignition_file.bootkube_sh: 1 error(s) occurred:

* module.assets_base.module.bootkube.data.ignition_file.bootkube_sh: data.ignition_file.bootkube_sh: unexpected EOF
```

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/151/pull-ci-origin-installer-e2e-aws/573/build-log.txt

wking commented:
> Looks like I may have broken the Terraform ignition setup?

Ah, I'd missed a `worker` -> `etcd` replacement in a copy/paste.  Fixed with bcb86ea -> 7fc19bd.

wking commented:
And we're [back to][1]:

> Ginkgo timed out waiting for all parallel nodes to report back!

Which is where we were before this PR, and distinct from the:

> Waiting for API at https://ci-op-ng25jj4j-68485-api.origin-ci-int-aws.dev.rhcloud.com:6443...

we [saw][2] during earlier versions of this PR (that just tried to open master -> etcd:10250).

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/157/pull-ci-origin-installer-e2e-aws/572/build-log.txt
[2]: https://github.com/openshift/installer/pull/151#issuecomment-414524652

wking commented:
I've fixed a:

```
* aws_security_group_rule.worker_ingress_flannel_from_worker: [WARN] A duplicate Security Group rule was found on (sg-008d3915f65c9e488). This may be
a side effect of a now-fixed Terraform issue causing two security groups with
identical attributes but different source_security_group_ids to overwrite each
other in the state. See https://github.com/hashicorp/terraform/pull/2376 for more
information and instructions for recovery. Error message: the specified rule “peer: sg-03a1212a84b765ddf, UDP, from port: 4789, to port: 4789, ALLOW” already exists
```

reported by @smarterclayton with 7fc19bd -> 81373c3.  Looks like I had accidentally pasted some worker stuff into the master rules.

wking commented:
And we're still [getting Ginkgo timeouts][1] with 81373c3fb8.

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/151/pull-ci-origin-installer-e2e-aws/582/build-log.txt

wking commented:
And we're still [getting Ginkgo timeouts][1] with cd83ddb.

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/151/pull-ci-origin-installer-e2e-aws/592/build-log.txt

smarterclayton commented:
The timeout issue is three separate things:

1. no connectivity between master and etcd nodes on kubelet ports (blocks exec, proxy, debug info gathered by e2e tests)
2. various selinux in RHCoS blockages - Derek is fixing in the operators repo by correctly making the pods privileged, but we need to be moving forward rather than back so I'd prefer not to revert to container linux
3. canal doesn't support host network to talk to pod network, so metrics-server can't be reached, which prevents namespaces from being terminated (an api group is down, namespace controller doesn't know whether there are any instances in the namespace so it blocks).  Casey said workaround is to disable all network policy in kube-system which I will try.  This is a really big blocker for aggregated apis so we need move to openshift-sdn and off canal as soon as possible.

wking commented:
> various selinux in RHCoS blockages - Derek is fixing in the operators repo by correctly making the pods privileged, but we need to be moving forward rather than back so I'd prefer not to revert to container linux

I agree about moving forward, but I'm fine going back to CoreOS for a few days to unblock ghe merge queue while he finishes that up.

> This is a really big blocker for aggregated apis so we need move to openshift-sdn and off canal as soon as possible.

This also sounds good.  But is there something we can land now to unblock the merge qurue?  With[these options][1], maybe that's switching from Calico to Flannel?

And I still don't understand what changed last week to suddenly surface these issues...

[1]: https://github.com/openshift/installer/blob/00066a96ddf772d68a31b7aba317ce7dee7160ca/examples/tectonic.aws.yaml#L253-L262

wking commented:
> With these options, maybe that's switching from Calico to Flannel?

I've switched the default to Flannel with 132e386 ->  b17dedc.

wking commented:
[The smoke test error][1] was:

```
18:24:51 * aws_launch_configuration.master_conf: InvalidAMIID.NotFound: The image id '[ami-ab6963d4]' does not exist
```

It's probably running in a region that's not us-east-1, so it's choking on 132e386.  I'll wait and see how the current e2e-aws tests go and then drop that commit.

wking commented:
Hooray, with b17dedc19c6a6aa5432675f5418c023962ae437d we're off the Ginkgo timeouts, and are only [getting][1]:

```
• Failure [34.901 seconds]
[Conformance][Area:Networking][Feature:Router]
/tmp/openshift/build-rpms/rpm/BUILD/origin-3.11.0/_output/local/go/src/github.com/openshift/origin/test/extended/router/headers.go:21
  The HAProxy router
  /tmp/openshift/build-rpms/rpm/BUILD/origin-3.11.0/_output/local/go/src/github.com/openshift/origin/test/extended/router/headers.go:40
    should set Forwarded headers appropriately [Suite:openshift/conformance/parallel] [It]
    /tmp/openshift/build-rpms/rpm/BUILD/origin-3.11.0/_output/local/go/src/github.com/openshift/origin/test/extended/router/headers.go:41

    Aug 23 18:38:47.491: Unexpected header: '10.2.6.0' (expected 10.0.192.68); All headers: http.Header{"X-Forwarded-Port":[]string{"8080"}, "X-Forwarded-Proto":[]string{"http"}, "Forwarded":[]string{"for=10.2.6.0;host=router-headers.example.com;proto=http"}, "X-Forwarded-For":[]string{"10.2.6.0"}, "User-Agent":[]string{"curl/7.61.0"}, "Accept":[]string{"*/*"}, "X-Forwarded-Host":[]string{"router-headers.example.com"}}

    /tmp/openshift/build-rpms/rpm/BUILD/origin-3.11.0/_output/local/go/src/github.com/openshift/origin/test/extended/router/headers.go:103
```

With b17dedc -> 2f1d81c, I've:

* Touched up the Calico -> Flannel commit to also update some fixtures,
* Returned to respecting user AMI overrides, and
* Rebased this on top of #134 to pick up the SELinux changes needed to work with new RHCOS AMIs.

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/151/pull-ci-origin-installer-e2e-aws/601/build-log.txt

wking commented:
I've pushed 2f1d81c -> b67f809 to leave the Calico fixtures alone.  They seem to be for internal unit tests unaffected by the default change or (for `tests/smoke/aws/vars/aws-basic.yaml`) they may be dead code (I'm dropping `tests/smoke/aws` in #143).

wking commented:
And we've [hit our quotas][1] in the Jenkins account...

```
Error: Error applying plan:

7 error(s) occurred:

* module.vpc.aws_nat_gateway.nat_gw[1]: 1 error(s) occurred:

* aws_nat_gateway.nat_gw.1: Error creating NAT Gateway: NatGatewayLimitExceeded: Performing this operation would exceed the limit of 5 NAT gateways
	status code: 400, request id: fe819d69-b3d5-44da-b71d-5d0113f5d7fd
* module.vpc.aws_elb.api_internal: 1 error(s) occurred:

* aws_elb.api_internal: TooManyLoadBalancers: Exceeded quota of account 846518947292
	status code: 400, request id: 34db82f2-a70b-11e8-b742-172fdc8dbc5d
* module.vpc.aws_elb.api_external: 1 error(s) occurred:

* aws_elb.api_external: TooManyLoadBalancers: Exceeded quota of account 846518947292
	status code: 400, request id: 34e016ce-a70b-11e8-ac02-dd9cffd48ed9
* module.vpc.aws_nat_gateway.nat_gw[0]: 1 error(s) occurred:

* aws_nat_gateway.nat_gw.0: Error creating NAT Gateway: NatGatewayLimitExceeded: Performing this operation would exceed the limit of 5 NAT gateways
	status code: 400, request id: a6ac1e8c-e85d-4578-8966-57c92850e5f9
* module.vpc.aws_nat_gateway.nat_gw[2]: 1 error(s) occurred:

* aws_nat_gateway.nat_gw.2: Error creating NAT Gateway: NatGatewayLimitExceeded: Performing this operation would exceed the limit of 5 NAT gateways
	status code: 400, request id: a409d983-ed38-4b80-83fe-1796e5f383c0
* module.vpc.aws_elb.tnc: 1 error(s) occurred:

* aws_elb.tnc: TooManyLoadBalancers: Exceeded quota of account 846518947292
	status code: 400, request id: 351fde8c-a70b-11e8-8aa8-937689137735
* module.vpc.aws_elb.console: 1 error(s) occurred:

* aws_elb.console: TooManyLoadBalancers: Exceeded quota of account 846518947292
	status code: 400, request id: 352695ff-a70b-11e8-85bf-3f8eb4d4beb0
```

This happened before [here][2].  I'll see about reaping leaked resources in that account.

[1]: https://jenkins-tectonic-installer.prod.coreos.systems/blue/rest/organizations/jenkins/pipelines/openshift-installer/branches/PR-151/runs/18/nodes/17/steps/26/log/?start=0
[2]: https://github.com/openshift/installer/pull/99#issuecomment-409982964

smarterclayton commented:
Yay!  This is an actual bug in the way that openshift/installer configures the router, because source ip isn't preserved.

I'll probably switch to another test in the job definition.

smarterclayton commented:
I switched to a kube conformance test that should pass and added a new job that runs all tests but is optional.

/retest

wking commented:
> I'll see about reaping leaked resources in that account.

I've removed a lot of cruft from our Jenkins smoke-test account.  Let's kick that off again:

retest this please

wking commented:
> I switched to a kube conformance test that should pass...

Cross-linking openshift/release#1271.

eparis commented:
/me hopes and hopes and hopes...

smarterclayton commented:
2018/08/23 21:12:28 Container test in pod e2e-aws completed successfully

!!!!

wking commented:
e2e-aws is green :) (although the smoke tests are still running).  Someone want to drop an /lgtm onto this?

eparis commented:
/lgtm
assuming I count

wking commented:
> assuming I count

I think all members of GitHub's openshift org count for `/lgtm`.

eparis commented:
Ohhh snap!  Thank you for sticking with this @wking
